src/aspire/source/micrograph.py

```diff
@@ -1,16 +1,225 @@
+        :return: `MicrographSource`
+        """
+
+        if (micrographs is None) or (
+            hasattr(micrographs, "__len__") and len(micrographs) == 0
+        ):
+            raise RuntimeError(
+                f"Must supply non-empty `micrographs` argument, received {micrographs}."
+            )
+
+        # Passed Array
+        if isinstance(micrographs, np.ndarray):
+            # Ensure dimensions
+            if micrographs.ndim == 2:
+                micrographs = micrographs[None, :, :]
+            elif micrographs.ndim != 3 or (
+                micrographs.shape[-2] != micrographs.shape[-1]
+            ):
+                raise NotImplementedError(
+                    f"Incompatible `micrographs` shape {micrographs.shape}, expects (count, L, L)"
+                )
+
+            self.micrograph_count = micrographs.shape[0]
+            self.micrograph_size = micrographs.shape[-1]
+            # Assign dtype override, or infer from `micrographs`.
+            self.dtype = dtype or micrographs.dtype
+
+            # We're already backed by an array, access it directly.
+            self._data = micrographs.astype(self.dtype, copy=False)
+            self._images_accessor = _ImageAccessor(
+                self._images_from_array, self.micrograph_count
+            )
+
+        elif self._is_path_like_input(micrographs):
+            if isinstance(micrographs, list):
+                # Explicit list
+                self.micrograph_files = micrographs
+            else:
+                # `str` cast will accommodate string and Path objects
+                self.micrograph_files = self._glob_files(str(micrographs))
+
+            # Assign count of micrograph files
+            self.micrograph_count = len(self.micrograph_files)
+
+            # Load the first micrograph to infer shape/type
+            # Size will be checked during on-the-fly loading of subsequent micrographs.
+            micrograph0 = Image.load(self.micrograph_files[0])
+
+            self.micrograph_size = micrograph0.resolution
+            # Assign dtype override, or infer from `micrograph0`.
+            self.dtype = dtype or micrograph0.dtype
+
+            # Prepare accessor to load files from disk on the fly.
+            self._images_accessor = _ImageAccessor(
+                self._images_from_files, self.micrograph_count
+            )
+        else:
+            raise NotImplementedError(
+                f"MicrographSource not implemented for {type(micrographs)}."
+            )
+
+    @staticmethod
+    def _is_path_like_input(p):
+        """
+        Utility to return whether `p` should be treated as a path to micrograph(s).
+
+        :param p: Item to test.
+        :return: Boolean.
+        """
+        return any([isinstance(p, list), isinstance(p, str), isinstance(p, Path)])
+
+    def __getitem__(self, key):
+        """
+        Getter utility function.
+        """
+        return self.__class__(self.images[key])
```

Comment:
Can we move a bracket in one of the tests and get coverage on this?

Reply:
meh, just deleted it. It's not used.